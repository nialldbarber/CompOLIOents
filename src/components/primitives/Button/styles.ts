import { CSSProp, css } from 'styled-components'
import styled from 'styled-components/native'
import { centerContent } from '~@theme/base/mixins'
import radius from '~@theme/base/radius'
import { TButtonVariants } from '~@theme/variants/button'
import type { TSize } from '~@theme/variants/sizes'

type TContainerProps = {
  $variants?: TButtonVariants
  $disabled?: boolean
  $styles?: CSSProp
  $sizes?: TSize
}

const disabledStyles = css`
  background-color: grey;
`

export const Container = styled.Pressable<TContainerProps>`
  border-radius: ${radius.sm};
  ${centerContent};
  ${({ $disabled }) => $disabled && disabledStyles};
  ${({ $styles }) => $styles && $styles};

  ${({ $sizes }) => $sizes && $sizes};
  ${({ $variants }) => $variants && $variants};
`
