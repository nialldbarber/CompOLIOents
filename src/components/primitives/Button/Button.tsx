import React from 'react'
import type { TouchableOpacityProps } from 'react-native'
import type { CSSProp } from 'styled-components'
import { Container } from '~@components/primitives/Button/styles'
import { Text } from '~@components/primitives/Text'
import { Factory } from '~@core/factory'
import hitSlop from '~@theme/base/hitSlop'
import type { THitSlop } from '~@theme/base/hitSlop'
import buttonVariants, { type TButtonVariants } from '~@theme/variants/button'
import sizeVariants, { type TSize } from '~@theme/variants/sizes'

/**
 * TODOS:
 * Find a way to add hitSlop back in
 * - currently pressable props don't recognise
 *   hitSlop as a prop
 */

interface TButtonProps extends TouchableOpacityProps {
  text?: string
  size?: TSize
  accessible?: boolean
  accessibilityLabel: string
  accessibilityHint?: string
  accessibilityLanguage?: string
  disabled?: boolean
  hitSlop?: THitSlop
  color?: TButtonVariants
  onPress?: (...args: any[]) => any
  children?: JSX.Element | Array<JSX.Element>
  styles?: CSSProp
  refs?: any
}

const Button = ({
  text,
  size = 'medium',
  accessible = true,
  accessibilityLabel,
  accessibilityHint,
  accessibilityLanguage,
  disabled,
  hitSlop = hitSlop,
  color = 'primary',
  onPress,
  children,
  styles,
  refs,
  ...rest
}: TButtonProps) => {
  const B = Factory(Container)
  const variants = buttonVariants[color]
  const sizes = sizeVariants[size]

  const buttonContent = text ? (
    <Text size={size} variants={color}>
      {text}
    </Text>
  ) : (
    children
  )

  return (
    <B
      ref={refs}
      accessible={accessible}
      accessibilityLabel={accessibilityLabel ?? text}
      accessibilityHint={accessibilityHint}
      accessibilityLanguage={accessibilityLanguage}
      disabled={disabled}
      onPress={onPress}
      $variants={variants}
      $sizes={sizes}
      $styles={styles}
      $disabled={disabled}
      {...rest}
    >
      {buttonContent}
    </B>
  )
}

export default Button
