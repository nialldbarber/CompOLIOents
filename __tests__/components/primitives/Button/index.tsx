import React from 'react'
import { fireEvent, render } from '@testing-library/react-native'
import { Text } from 'react-native'
import { Button } from '~@components/primitives/Button'

describe('<Button />', () => {
  const onPress = jest.fn()
  const accessibilityLabel = 'hello'
  const props = {
    text: 'button test',
    onPress,
    accessibilityLabel,
  }

  it('should match the snapshot', () => {
    const { toJSON } = render(<Button {...props} />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('should call the provided function when pressed', () => {
    const { getByText } = render(<Button {...props} />)
    const button = getByText(/bUttOn tEsT/i)
    fireEvent.press(button)
    expect(onPress).toHaveBeenCalled()
  })

  it('should render children instead of text if provided', () => {
    const { toJSON } = render(
      <Button accessibilityLabel={accessibilityLabel}>
        <Text>Push it to the limit</Text>
      </Button>
    )
    expect(toJSON()).toMatchSnapshot()
  })
})
