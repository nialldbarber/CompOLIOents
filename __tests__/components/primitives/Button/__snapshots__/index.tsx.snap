// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match the snapshot 1`] = `
<View
  accessibilityLabel="hello"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "dodgerblue",
        "borderColor": "dodgerblue",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 1,
        "color": "#111",
        "display": "flex",
        "fontSize": 16,
        "justifyContent": "center",
        "paddingHorizontal": 12,
        "paddingVertical": 12,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#111",
          "fontSize": 16,
          "fontStyle": "normal",
          "fontWeight": "normal",
          "textDecorationColor": "black",
          "textDecorationLine": "none",
          "textDecorationStyle": "solid",
        },
      ]
    }
  >
    button test
  </Text>
</View>
`;

exports[`<Button /> should render children instead of text if provided 1`] = `
<View
  accessibilityLabel="hello"
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "dodgerblue",
        "borderColor": "dodgerblue",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 1,
        "color": "#111",
        "display": "flex",
        "fontSize": 16,
        "justifyContent": "center",
        "paddingHorizontal": 12,
        "paddingVertical": 12,
      },
    ]
  }
>
  <Text>
    Push it to the limit
  </Text>
</View>
`;
